File: apps/mobile/test/first_run_locale_test.dart
// U1 — the one-time first-run language question (phone). Mechanism tests.
//
// What this file has to prove, in the card's own words:
//   ① the picker FIRES for a new install and NEVER for an existing user (a
//      never-fires façade and a nag loop are both failures);
//   ② the answer goes to the SAME setting the settings row writes, and takes
//      effect immediately;
//   ③ 🔴 the no-OS-locale red line still holds — asking a human once is not
//      inferring from the environment.
//
// ⚠️ THE MEASUREMENT THIS FILE IS BUILT ON (stated so the next reader does not
// have to re-derive it): neither `flowmic.pref.locale` nor its desktop twin is
// ever written eagerly — `setLocale` is its only writer and it runs on an
// explicit user change. So「the language key is missing」is true for a brand-new
// install AND for every existing user who never opened the language row: those
// two populations are byte-for-byte identical on that signal, and they must go
// opposite ways. The witness that separates them is the REST of the profile,
// and the key used below to stand for「an existing user」is the real one:
// `flowmic.timeline.migrated.sqlite.v1`, which timeline_sqlite.dart stamps on
// the first boot of every install — including a fresh one, which is exactly why
// the resolution has to happen in load(), before that write.

import 'dart:io';

import 'package:flowmic/src/settings/app_settings.dart';
import 'package:flowmic/src/settings/app_strings.dart';
import 'package:flowmic/src/ui/first_run_locale_page.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'support/legibility.dart';

/// The witness of an install that has booted before — the real key, not a
/// stand-in. See the file header.
const String kExistingUserWitness = 'flowmic.timeline.migrated.sqlite.v1';
const String kPromptKey = 'flowmic.pref.locale_prompt';
const String kLocaleKey = 'flowmic.pref.locale';

/// One "boot": a fresh controller over the prefs as they stand right now.
Future<AppSettingsController> boot(SharedPreferences prefs) async {
  final AppSettingsController c = AppSettingsController(prefs: prefs);
  await c.load();
  return c;
}

Future<SharedPreferences> prefsWith(Map<String, Object> initial) async {
  SharedPreferences.setMockInitialValues(initial);
  return SharedPreferences.getInstance();
}

void main() {
  group('U1 — when the first-run question may be asked', () {
    test('NEW INSTALL: nothing stored at all → asks, and the ask is recorded',
        () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      expect(c.needsLocaleChoice, isTrue);
      expect(prefs.getString(kPromptKey), AppSettingsController.kPromptPending);
      // The app behind the question is still the documented default — the
      // picker does not pre-answer anything.
      expect(c.locale, AppLocale.en);
      expect(prefs.getString(kLocaleKey), isNull);
    });

    test('🔴 EXISTING USER who never opened the language row → never asked, '
        'and their language is left exactly as it was', () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{
        kExistingUserWitness: true,
      });
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      expect(c.needsLocaleChoice, isFalse);
      expect(prefs.getString(kPromptKey), AppSettingsController.kPromptSettled);
      // Grandfathering writes the PROMPT key, never the LOCALE key: the status
      // quo (implicit default, now English) is preserved, not re-declared as a
      // choice they never made.
      expect(prefs.getString(kLocaleKey), isNull);
      expect(c.locale, AppLocale.en);
    });

    test('EXISTING USER with an explicit choice → never asked, choice kept',
        () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{
        kExistingUserWitness: true,
        kLocaleKey: 'ja',
      });
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      expect(c.needsLocaleChoice, isFalse);
      expect(c.locale, AppLocale.ja);
    });

    test('any other flowmic pref is enough of a witness (send policy alone)',
        () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{
        'flowmic.compose.send_policy': 'manual',
      });
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      expect(c.needsLocaleChoice, isFalse);
    });

    test('the marker itself is not mistaken for prior state', () async {
      // Otherwise the very act of recording「we asked」would make the next boot
      // read the profile as an upgrade and drop the question on the floor.
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController first = await boot(prefs);
      addTearDown(first.dispose);
      expect(first.needsLocaleChoice, isTrue);
      // Nothing else was written; only the marker exists now.
      expect(prefs.getKeys(), <String>{kPromptKey});
      final AppSettingsController second = await boot(prefs);
      addTearDown(second.dispose);
      expect(second.needsLocaleChoice, isTrue);
    });

    test('asked ONCE: a settled install is never asked again, however many '
        'boots', () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController first = await boot(prefs);
      addTearDown(first.dispose);
      first.chooseLocale(AppLocale.ko);
      for (int i = 0; i < 3; i++) {
        final AppSettingsController later = await boot(prefs);
        addTearDown(later.dispose);
        expect(later.needsLocaleChoice, isFalse, reason: 'boot #${i + 2}');
        expect(later.locale, AppLocale.ko);
      }
    });

    test('killed mid-question → asked again next boot (pending ≠ settled)',
        () async {
      // Without this branch: first boot shows the picker → the user kills the
      // app → that boot has already stamped the timeline flag → the next boot
      // reads the profile as an upgrade → a new install is stranded in a
      // language nobody chose.
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController first = await boot(prefs);
      addTearDown(first.dispose);
      expect(first.needsLocaleChoice, isTrue);
      await prefs.setBool(kExistingUserWitness, true); // the boot got this far
      final AppSettingsController second = await boot(prefs);
      addTearDown(second.dispose);
      expect(second.needsLocaleChoice, isTrue);
    });
  });

  group('U1 — the answer is the same setting the settings row writes', () {
    test('chooseLocale writes flowmic.pref.locale — the same key setLocale does',
        () async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      c.chooseLocale(AppLocale.en);
      expect(prefs.getString(kLocaleKey), 'en');
      expect(prefs.getString(kPromptKey), AppSettingsController.kPromptSettled);
      expect(c.locale, AppLocale.en);
      expect(c.needsLocaleChoice, isFalse);
      // Positive control: the settings row's writer lands in the same slot, so
      // there is exactly one answer to「which language is the UI」and the two surfaces can
      // never disagree.
      c.setLocale(AppLocale.ja);
      expect(prefs.getString(kLocaleKey), 'ja');
      final AppSettingsController next = await boot(prefs);
      addTearDown(next.dispose);
      expect(next.locale, AppLocale.ja);
      expect(next.needsLocaleChoice, isFalse);
    });

    test('🔴 choosing English — the DEFAULT — still settles the question',
        () async {
      // The trap this second key exists for: setLocale returns early when the
      // value is unchanged, so picking the default writes NOTHING through it.
      // A picker gated on「is the language key present」would come back forever.
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      expect(c.needsLocaleChoice, isTrue);
      c.chooseLocale(AppLocale.en);
      expect(c.needsLocaleChoice, isFalse);
      final AppSettingsController next = await boot(prefs);
      addTearDown(next.dispose);
      expect(next.needsLocaleChoice, isFalse);
      expect(next.locale, AppLocale.en);
    });
  });

  group('U1 — what the user actually sees', () {
    testWidgets('renders all four languages in their own script, with nothing '
        'pre-selected, and nothing clipped on a 360dp phone',
        (WidgetTester tester) async {
      tester.view.physicalSize = const Size(360, 720);
      tester.view.devicePixelRatio = 1.0;
      addTearDown(tester.view.resetPhysicalSize);
      addTearDown(tester.view.resetDevicePixelRatio);

      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);
      await tester.pumpWidget(
        MaterialApp(home: FirstRunLocalePage(appSettings: c)),
      );

      // The endonyms — the point of the screen: readable without already
      // knowing the UI language.
      for (final String name in <String>['中文', 'English', '日本語', '한국어']) {
        expect(find.text(name), findsOneWidget, reason: name);
      }
      // …and the word「language」in each of the four, so the screen says what it
      // is asking in every language it offers. Straight from the existing
      // catalogue — no new key was invented for this screen.
      for (final AppLocale l in AppLocale.values) {
        expect(find.text(AppStrings.of(l).uiLanguage), findsOneWidget);
      }

      // 🔴 RENDERED RESULT, not Text.data: the readability claim is「the user
      // can read the option」, and 15 册's lesson is that a Text whose data is
      // right can still be three dots on screen. Assert on the laid-out box.
      //
      // 🔴 CORRECTED 2026-08-07 (W5a adversarial review P1-1, measured). What
      // stood here asserted the right INTENT with two readings that cannot fail:
      //   · `didExceedMaxLines` — `first_run_locale_page.dart` never sets
      //     `maxLines` (grep: zero hits), and that getter is always false
      //     without one;
      //   · `size.width <= constraints.maxWidth` — `RenderParagraph` sets
      //     `size = constraints.constrain(textSize)`, so that inequality holds
      //     BY CONSTRUCTION, for every paragraph, always.
      // Two assertions, both permanently green, guarding nothing. They now go
      // through `support/legibility.dart`, which picks its criterion from the
      // paragraph's own structure — and `size.width > 0` is kept, because a
      // paragraph that rendered nothing would satisfy anything else here.
      for (final String name in <String>['中文', 'English', '日本語', '한국어']) {
        final Finder f = find.text(name);
        expectLegible(tester, f, reason: name);
        expect(tester.getSize(f).width, greaterThan(0), reason: name);
      }
      expect(tester.takeException(), isNull);
    });

    testWidgets('a tap IS the choice: it writes the one language key and the '
        'app is in that language on the next frame — no restart',
        (WidgetTester tester) async {
      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);

      // A host shaped like main.dart's gate: the picker while the question is
      // owed, the app afterwards, rebuilt by the controller's own notification.
      await tester.pumpWidget(
        MaterialApp(
          home: ListenableBuilder(
            listenable: c,
            builder: (BuildContext context, Widget? child) =>
                c.needsLocaleChoice
                    ? FirstRunLocalePage(appSettings: c)
                    : Scaffold(
                        body: Center(
                          child: Text(AppStrings.of(c.locale).settingsTitle),
                        ),
                      ),
          ),
        ),
      );
      expect(find.text('한국어'), findsOneWidget);

      await tester.tap(find.text('한국어'));
      await tester.pumpAndSettle();

      // The setting: the same key the settings row writes.
      expect(prefs.getString(kLocaleKey), 'ko');
      expect(prefs.getString(kPromptKey), AppSettingsController.kPromptSettled);
      // Immediately: the question is gone and the app behind it is Korean.
      expect(find.text('한국어'), findsNothing);
      expect(find.text('설정'), findsOneWidget); // ko settingsTitle
      expect(find.text('设置'), findsNothing); // …and not the zh default
    });
  });

  group('U1 — 🔴 the no-OS-locale red line still holds', () {
    testWidgets('a ja/zh OS with nothing stored gets ASKED, not answered: the '
        'UI behind the question is still the English default and nothing is pre-selected',
        (WidgetTester tester) async {
      // The trap cases: an OS whose language the app DOES support, and that is
      // NOT the default. Stubbing to en-US after the default flipped to English
      // would go green even if the code started reading the platform locale.
      tester.platformDispatcher.localeTestValue = const Locale('ja', 'JP');
      addTearDown(tester.platformDispatcher.clearLocaleTestValue);

      final SharedPreferences prefs = await prefsWith(<String, Object>{});
      final AppSettingsController c = await boot(prefs);
      addTearDown(c.dispose);

      expect(c.needsLocaleChoice, isTrue); // asked
      expect(c.locale, AppLocale.en); // not answered for them
      expect(prefs.getString(kLocaleKey), isNull);

      await tester.pumpWidget(
        MaterialApp(home: FirstRunLocalePage(appSettings: c)),
      );
      // All four rows are offered flat — none is marked as the OS's.
      for (final String name in <String>['中文', 'English', '日本語', '한국어']) {
        expect(find.text(name), findsOneWidget);
      }

      // A mid-session OS flip changes nothing either: there is no OS-locale
      // listener anywhere in the language path (the THEME may follow the OS;
      // the language may NOT — two deliberately different rulings).
      tester.platformDispatcher.localeTestValue = const Locale('zh', 'CN');
      await tester.pumpAndSettle();
      expect(c.locale, AppLocale.en);
      expect(c.needsLocaleChoice, isTrue);
    });

    test('structural: the first-run path contains no platform-locale probe',
        () async {
      // The runtime case above proves today; this anchors tomorrow — a future
      // 「helpfully pre-select the row matching the phone's language」 turns this
      // red before it can ship. Same guard shape as the desktop's
      // first-run-locale.test.ts.
      //
      // 🔴 P-7 (W5a): THIS LIST IS THE GUARD'S ENTIRE FIELD OF VIEW, so anything
      // that JOINS the first-run path has to join the list on the same day.
      // Otherwise the guard keeps passing while the thing it guards grows — a
      // green check whose coverage silently shrank, which reads exactly like a
      // green check whose subject is safe. The onboarding screens below now run
      // between the language question and the instance list; they are prose
      // about how to use the app, which is precisely the kind of screen somebody
      // would later be tempted to 「helpfully」 localise off the OS.
      const List<String> sources = <String>[
        'lib/src/ui/first_run_locale_page.dart',
        'lib/src/settings/app_settings.dart',
        'lib/src/ui/onboarding/first_run_onboarding_page.dart',
        'lib/src/ui/onboarding/onboarding_view.dart',
        'lib/src/ui/onboarding/onboarding_art.dart',
      ];
      for (final String path in sources) {
        final String src = await File(path).readAsString();
        expect(src, isNot(contains('PlatformDispatcher.instance.locale')),
            reason: path);
        expect(src, isNot(contains('Platform.localeName')), reason: path);
        expect(src, isNot(contains('Localizations.localeOf')), reason: path);
        expect(src, isNot(contains('window.locale')), reason: path);
      }
    });
  });
}

File: apps/desktop/src/lib/strings/first-run-locale.test.ts
// U1 first-run language gate — mechanism tests.
//
// What this file has to prove, in the card's own words:
//   ① the picker FIRES for a new install and NEVER for an existing user (the
//      never-fires façade and the nag loop are both failures);
//   ② the answer goes to the SAME setting the switcher writes, and takes
//      effect immediately;
//   ③ 🔴 the no-OS-locale red line still holds — asking a human once is not
//      inferring from the environment, and nothing here reads the environment.
//
// The red-line pair mirrors locale.test.ts's: one RUNTIME case (stub the OS to
// a shipped language that is NOT the default — ja-JP — keep the profile empty,
// assert the app is still DEFAULT en behind the question) and one STRUCTURAL
// case (the sources contain no OS probe at all), because the runtime one proves
// today and the structural one anchors tomorrow.
// ⚠️ Stubbing the OS to en-US after the default flipped to English would go
// green even if the code started reading navigator.language.

import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';
import { createSSRApp } from 'vue';
import { renderToString } from 'vue/server-renderer';
import { afterEach, describe, expect, it, vi } from 'vitest';
import type { KvStore } from '../types';
import {
  DEFAULT_LOCALE,
  LOCALE_ENDONYM,
  LOCALE_KEY,
  UI_LOCALES,
  getLocale,
  hydrateLocale,
  setLocale,
  wireLocaleStore,
} from './locale';
import {
  LOCALE_PROMPT_KEY,
  PROMPT_PENDING,
  PROMPT_SETTLED,
  chooseFirstRunLocale,
  hasPriorFlowMicState,
  resolveFirstRunPrompt,
  settleLocalePrompt,
} from './first-run-locale';
import { S, S_BY_LOCALE } from '../strings';
import FirstRunLocale from '../../main-window/components/FirstRunLocale.vue';

function memKv(initial?: Record<string, string>): KvStore & { map: Map<string, string> } {
  const map = new Map(Object.entries(initial ?? {}));
  return {
    map,
    get: (k) => map.get(k) ?? null,
    set: (k, v) => void map.set(k, v),
  };
}

/** What a boot stamps into a BRAND-NEW profile before any user has done
 *  anything — measured, not assumed: main-window/store.ts constructs
 *  TimelineStore at module scope and its constructor ends in `persist()`, which
 *  writes these three (rows / image ids / retention cutoffs = NO_CUTOFFS). ES
 *  imports run before every statement, so the gate can never precede them. */
const BOOT_STAMPS: Record<string, string> = {
  'flowmic.history.cache': '[]',
  'flowmic.history.images': '[]',
  'flowmic.history.retention': '{"text":null,"images":null}',
};

/** One "boot": the profile's keys are whatever the store already holds, which
 *  is the same thing App.vue reads out of localStorage. */
function boot(kv: KvStore & { map: Map<string, string> }): boolean {
  wireLocaleStore(kv);
  hydrateLocale();
  return resolveFirstRunPrompt(kv, [...kv.map.keys()]);
}

/** A boot of the REAL app: the timeline's module-scope stamps land first. */
function bootWithTimelineStamps(kv: KvStore & { map: Map<string, string> }): boolean {
  for (const [k, v] of Object.entries(BOOT_STAMPS)) kv.set(k, v);
  return boot(kv);
}

afterEach(() => {
  vi.unstubAllGlobals();
  wireLocaleStore(null);
  setLocale(DEFAULT_LOCALE);
});

describe('U1 — when the first-run question may be asked', () => {
  it('NEW INSTALL: nothing stored at all → asks, and the ask is recorded', () => {
    const kv = memKv();
    expect(boot(kv)).toBe(true);
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_PENDING);
  });

  it('🔴 NEW INSTALL still asks after the timeline stamps its three keys — the '
    + 'measured hazard that makes this a real façade risk', () => {
    // A presence-only witness fails RIGHT HERE, and only here: store.ts builds
    // TimelineStore at module scope, so a brand-new profile is already three
    // keys deep before any statement in App.vue runs. The gate asks whether a
    // key holds SOMETHING, and empty containers hold nothing.
    const kv = memKv();
    expect(bootWithTimelineStamps(kv)).toBe(true);
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_PENDING);
  });

  it('…but a timeline with ROWS in it is an existing user', () => {
    // The same three keys, one of them actually filled: this machine has
    // received sentences, so it is not a first run whatever the marker says.
    const kv = memKv({ 'flowmic.history.cache': '[{"id":"r1"}]' });
    expect(boot(kv)).toBe(false);
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_SETTLED);
  });

  it('🔴 EXISTING USER who never opened the switcher → never asked, and their '
    + 'language is left exactly as it was', () => {
    // The measurement behind this case: NEITHER end writes the locale eagerly
    // (locale.ts setLocale is its only writer and it runs on an explicit
    // choice), so this user and a new install are byte-for-byte identical on
    // 「is flowmic.ui.locale missing」. The witness that separates them is the
    // rest of their profile — here, a theme they once picked.
    const kv = memKv({ 'flowmic.ui.theme': 'dark' });
    expect(boot(kv)).toBe(false);
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_SETTLED);
    // Grandfathering writes the PROMPT key, never the LOCALE key: their status
    // quo (DEFAULT_LOCALE, now en) is preserved rather than re-declared as a choice.
    expect(kv.get(LOCALE_KEY)).toBeNull();
    expect(getLocale()).toBe(DEFAULT_LOCALE);
  });

  it('EXISTING USER with an explicit choice → never asked, choice kept', () => {
    const kv = memKv({ [LOCALE_KEY]: 'ja' });
    expect(boot(kv)).toBe(false);
    expect(getLocale()).toBe('ja');
  });

  it('the marker itself is not mistaken for prior state', () => {
    // Otherwise the very act of recording「we asked」would make the next boot
    // read the profile as an upgrade.
    const kv = memKv();
    expect(hasPriorFlowMicState([LOCALE_PROMPT_KEY], kv)).toBe(false);
    const withRows = memKv({ 'flowmic.history.cache': '[{"id":"r1"}]' });
    expect(hasPriorFlowMicState([LOCALE_PROMPT_KEY, 'flowmic.history.cache'], withRows))
      .toBe(true);
    // Someone else's key on the same origin is not our history.
    const foreign = memKv({ theme: 'dark', 'sentry.session': '{"a":1}' });
    expect(hasPriorFlowMicState(['theme', 'sentry.session'], foreign)).toBe(false);
  });

  it('asked ONCE: a settled install is never asked again, however many boots', () => {
    const kv = memKv();
    expect(boot(kv)).toBe(true);
    chooseFirstRunLocale('ko', kv);
    expect(boot(kv)).toBe(false);
    expect(boot(kv)).toBe(false);
    expect(boot(kv)).toBe(false);
  });

  it('killed mid-question → asked again next boot (pending is not settled)', () => {
    // Without this branch the sequence 「first boot shows the picker → user
    // closes the window → boot writes its own keys → next boot reads the
    // profile as an upgrade」 strands a new install in a language nobody chose.
    const kv = memKv();
    expect(boot(kv)).toBe(true);
    kv.set('flowmic.history.cache', '[{"id":"r1"}]'); // this boot got as far as a row
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_PENDING);
    expect(boot(kv)).toBe(true);
  });

  it('storage unreadable → does not ask (a nag loop is the worse failure)', () => {
    const kv = memKv();
    expect(resolveFirstRunPrompt(kv, null)).toBe(false);
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_SETTLED);
  });
});

describe('U1 — the answer is the same setting the switcher writes', () => {
  it('choosing writes flowmic.ui.locale and swaps the catalogue immediately', () => {
    const kv = memKv();
    expect(boot(kv)).toBe(true);
    expect(S.nav_devices).toBe('Devices'); // English UI behind the question (fresh-install default)
    chooseFirstRunLocale('zh-CN', kv);
    // ONE store holds both keys — the language and「asked already」cannot drift.
    expect(kv.get(LOCALE_KEY)).toBe('zh-CN');
    expect(kv.get(LOCALE_PROMPT_KEY)).toBe(PROMPT_SETTLED);
    // Immediately, with no reload: the reactive S is already Chinese.
    expect(getLocale()).toBe('zh-CN');
    expect(S.nav_devices).toBe('设备');
    expect(S.nav_settings).toBe('设置');
  });

  it('🔴 choosing English — the DEFAULT — still settles the question', () => {
    // The trap: if「settled」were read off the locale key, picking the value
    // that happens to be the default would look like「never chose」and the
    // picker would come back on every boot.
    const kv = memKv();
    expect(boot(kv)).toBe(true);
    chooseFirstRunLocale('en', kv);
    expect(boot(kv)).toBe(false);
  });

  it('the switcher and the picker are interchangeable writers of one key', () => {
    const kv = memKv();
    boot(kv);
    chooseFirstRunLocale('ja', kv);
    expect(kv.get(LOCALE_KEY)).toBe('ja');
    setLocale('ko'); // what PrefsAppearance.vue does
    expect(kv.get(LOCALE_KEY)).toBe('ko');
    expect(boot(kv)).toBe(false);
    expect(getLocale()).toBe('ko');
  });

  it('settleLocalePrompt alone never touches the language', () => {
    const kv = memKv();
    boot(kv);
    settleLocalePrompt(kv);
    expect(kv.get(LOCALE_KEY)).toBeNull();
  });
});

describe('U1 — 🔴 the no-OS-locale red line still holds', () => {
  it('runtime: OS stubbed to ja-JP, brand-new profile → app is STILL DEFAULT en and '
    + 'the question is asked instead of answered for the user', async () => {
    vi.resetModules();
    vi.stubGlobal('navigator', { language: 'ja-JP', languages: ['ja-JP', 'ja'] });
    try {
      const freshLocale = await import('./locale');
      const freshGate = await import('./first-run-locale');
      const freshStrings = await import('../strings');
      const kv = memKv();
      freshLocale.wireLocaleStore(kv);
      expect(freshLocale.hydrateLocale()).toBe('en');
      // The gate ASKS — it does not answer. An OS-derived pre-answer would show
      // up right here as `false` plus a stored 'ja'.
      expect(freshGate.resolveFirstRunPrompt(kv, [...kv.map.keys()])).toBe(true);
      expect(kv.get(freshLocale.LOCALE_KEY)).toBeNull();
      expect(freshLocale.getLocale()).toBe('en');
      expect(freshStrings.S.nav_devices).toBe('Devices');
    } finally {
      vi.unstubAllGlobals();
      vi.resetModules();
    }
  });

  it('structural: neither the gate nor the picker contains an OS-locale probe', () => {
    // Same guard locale.test.ts puts on locale.ts. It has to cover the picker
    // too: 「helpfully pre-select the row matching navigator.language」 breaks
    // the red line just as thoroughly as reading it in the state module.
    for (const rel of ['./first-run-locale.ts', '../../main-window/components/FirstRunLocale.vue']) {
      const src = readFileSync(fileURLToPath(new URL(rel, import.meta.url)), 'utf8');
      expect(src, rel).not.toMatch(/navigator\./);
      expect(src, rel).not.toMatch(/Intl\./);
      expect(src, rel).not.toMatch(/matchMedia/);
    }
  });
});

describe('U1 — what the user actually sees', () => {
  it('renders every shipped language in its own script, with no pre-selection', async () => {
    wireLocaleStore(memKv());
    hydrateLocale();
    const html = await renderToString(createSSRApp(FirstRunLocale));
    // 🔴 DERIVED FROM THE REGISTRY, NOT THE FOUR NAMES THIS USED TO LIST
    // (2026-08-14). The old form asserted 中文 / English / 日本語 / 한국어 by hand,
    // which means a fifth language could ship missing from this screen with the
    // test green — and this screen is the ONE place a user who cannot read the
    // current UI language can fix that. Now the assertion grows with the list.
    for (const l of UI_LOCALES) {
      // The endonym — the point of the screen: readable without already knowing
      // the UI language.
      expect(html, `${l}: endonym`).toContain(LOCALE_ENDONYM[l]);
      // …and the word「language」in that language, so the screen says what it is
      // asking in every language it offers. (A language that has not translated
      // this key yet inherits the English word, which is the declared fallback —
      // still a real answer, not a blank.)
      expect(html, `${l}: the word for "language"`).toContain(S_BY_LOCALE[l].set_prefs_language);
      // One option per shipped locale, each individually clickable.
      expect(html, `${l}: option`).toContain(`data-locale="${l}"`);
    }
    // Measure the ruler (§1-bis-5): the loop above proves every registry row is
    // present; this proves the screen offers NOTHING ELSE — an extra row would
    // be a language the rest of the app does not have.
    expect((html.match(/data-locale="/g) ?? []).length).toBe(UI_LOCALES.length);
    // No「selected / current」affordance: with nothing chosen yet, highlighting
    // a row would be the app answering its own question.
    expect(html).not.toMatch(/class="[^"]*\bon\b[^"]*"/);
    expect(html).not.toContain('checked');
  });

  it('the screen renders identically whatever the app is currently rendered in', async () => {
    // Endonyms do not translate: the same rows greet a zh-CN install and an en
    // one, so this screen never becomes a second place where「what language is
    // this app」has to be answered before it can be read.
    wireLocaleStore(memKv());
    hydrateLocale();
    const zh = await renderToString(createSSRApp(FirstRunLocale));
    setLocale('en');
    const en = await renderToString(createSSRApp(FirstRunLocale));
    expect(en).toBe(zh);
  });

  it('the click is wired to the one writer and the one mirror', () => {
    // Literal anchor (the culture prefs-appearance.test.ts states for the same
    // wiring): the Tauri emit cannot be exercised in a node test, so what is
    // pinned is that the component calls the shared writer and the shared
    // prefs-sync broadcast rather than growing its own.
    const src = readFileSync(
      fileURLToPath(new URL('../../main-window/components/FirstRunLocale.vue', import.meta.url)),
      'utf8',
    );
    expect(src).toContain('chooseFirstRunLocale(l, localKv)');
    expect(src).toContain('notifyPrefsChanged()');
    // No second language key, and no writing localStorage behind the store.
    expect(src).not.toMatch(/localStorage/);
    expect(src).not.toMatch(/flowmic\.ui\.locale/);
  });
});
